### DISCUSSION TITLE
How to make reply keyboard button translatable ? (Laravel)

### DISCUSSION THREAD

**Discussion by ifteakhor** [nutgram/nutgram#747] [Q&A]
I have a bot and I want to make it multi language, but it doesn't work 😢 

![Screenshot 2024-07-15 at 17 42 04](https://github.com/user-attachments/assets/509bf580-69c8-4188-85a2-193b5391efdd)

I've added translation but in these file (`routes/telegram.php`) translation are always in English and bot can't respond to keyboards

<img width="652" alt="Screenshot 2024-07-15 at 17 41 45" src="https://github.com/user-attachments/assets/591715a3-8071-431e-b705-e47802305045">

I am setting user language in Nutgram middleware 

![Screenshot 2024-07-15 at 17 42 32](https://github.com/user-attachments/assets/4ac77aac-2c70-4819-b9d3-d55ed21702f3)

![Screenshot 2024-07-15 at 17 42 50](https://github.com/user-attachments/assets/0a6c0bb0-2b6a-488b-8169-aa77f5371d2e)

Any idea ? I tried to set language in `AppServiceProvider` but nutgram routes loads at first

**Comment by Lukasss93** [comment]
You're right: Nutgram routes are resolved at application startup.
Just do this:
```php
class LaravelTrans
{
    public function __construct(protected string $key) {
    }

    public function __invoke(Nutgram $bot, $next)
    {
        if($bot->message()?->text !== trans($this->key)){
            return;
        }
        
        $next($bot);
    }
}
```
```php
$bot->onText('.*', KeyboardButtonA::class)->middleware(new LaravelTrans('button.a'));
$bot->onText('.*', KeyboardButtonB::class)->middleware(new LaravelTrans('button.b'));
$bot->onText('.*', KeyboardButtonC::class)->middleware(new LaravelTrans('button.c'));
```

**Reply by ifteakhor** [reply]
` [try] TypeError => SergiX44\Nutgram\Nutgram::invoke(): Argument #1 ($callable) must be of type callable|array|string, App\Telegram\Middleware\LaravelTransMiddleware given, called in /home/tuner/web/marsu.toza.su/public_html/vendor/nutgram/nutgram/src/Middleware/Link.php on line 37`

**Reply by Lukasss93** [reply]
ops, post updated. changed `handle` with `__invoke`

**Reply by ifteakhor** [reply]
Yes, I've changed it but doesn't work

**Reply by ifteakhor** [reply]
<img width="831" alt="Screenshot 2024-07-16 at 00 11 01" src="https://github.com/user-attachments/assets/8470880d-4fdf-486b-82eb-75c2b7aee3ef">

I've added log here

**Reply by ifteakhor** [reply]
And this is what I am getting, it seems that Nutgram caches it?

<img width="1277" alt="Screenshot 2024-07-16 at 00 11 31" src="https://github.com/user-attachments/assets/2de89ad6-dab8-49e6-adc4-d731cea60b45">

<img width="759" alt="Screenshot 2024-07-16 at 00 12 06" src="https://github.com/user-attachments/assets/a753ff66-d9f5-4a7d-a353-43efb6ace128">


Keys are wrong, does Nutgram cache the key? It always gets `get_access` key

**Reply by ifteakhor** [reply]
Oh I get it I think. When nutgram loads the telegram routes file it set the `LaravelTranMiddleware` to `common.get_access`, because it is the last key

**Comment by Lukasss93** [comment]
nvm, you need to use `onMessage` handler instead of onText because the `onText` pattern must be unique.
remove the return in `$next($bot);` too

![immagine](https://github.com/user-attachments/assets/cf6d21e0-6be5-4f02-8faf-84105609c29a)

**Reply by ifteakhor** [reply]
I've fixed it this way, which way is better?

<img width="768" alt="Screenshot 2024-07-16 at 00 30 47" src="https://github.com/user-attachments/assets/5b8873ee-a62d-4ac8-b335-90fe19d13a5c">
